File: src/streaming.py
"""
Streaming Response - потоковая передача ответов от LLM.
Улучшает UX показывая ответ по мере генерации.
"""
from __future__ import annotations

import sys
import time
from typing import Iterator, Callable

from .llm_client import LLMClient, LLMMessage, LLMResponse


class StreamingHandler:
    """Обработчик потоковых ответов."""

    def __init__(self, on_token: Callable[[str], None] | None = None):
        """
        Args:
            on_token: Callback для каждого токена
        """
        self.on_token = on_token or self._default_token_handler
        self.buffer = ""
        self.total_tokens = 0

    def _default_token_handler(self, token: str) -> None:
        """Обработчик по умолчанию - вывод в stdout."""
        print(token, end="", flush=True)

    def handle_token(self, token: str) -> None:
        """
        Обработать токен.

        Args:
            token: Токен от LLM
        """
        self.buffer += token
        self.total_tokens += 1
        self.on_token(token)

    def get_full_response(self) -> str:
        """Получить полный ответ."""
        return self.buffer

    def reset(self) -> None:
        """Сбросить буфер."""
        self.buffer = ""
        self.total_tokens = 0


class StreamingLLMClient:
    """Обертка для LLM клиента с поддержкой streaming."""

    def __init__(self, llm_client: LLMClient):
        self.llm_client = llm_client

    def complete_streaming(
        self,
        messages: list[LLMMessage],
        system: str | None = None,
        on_token: Callable[[str], None] | None = None
    ) -> LLMResponse:
        """
        Получить ответ с потоковой передачей.

        Args:
            messages: История сообщений
            system: Системный промпт
            on_token: Callback для каждого токена

        Returns:
            LLMResponse с полным ответом
        """
        handler = StreamingHandler(on_token)

        # Проверяем поддерживает ли провайдер streaming
        if hasattr(self.llm_client, 'supports_streaming') and self.llm_client.supports_streaming():
            # Используем нативный streaming
            try:
                for token in self._stream_native(messages, system):
                    handler.handle_token(token)
            except Exception as e:
                # Fallback на обычный режим
                print(f"\n⚠️  Streaming failed, falling back to normal mode: {e}")
                response = self.llm_client.complete(messages, system)
                handler.buffer = response.content
                return response
        else:
            # Эмулируем streaming
            response = self.llm_client.complete(messages, system)
            self._emulate_streaming(response.content, handler)

        # Формируем финальный ответ
        return LLMResponse(
            content=handler.get_full_response(),
            model=self.llm_client.config.model,
            usage={
                "completion_tokens": handler.total_tokens,
                "prompt_tokens": 0,
                "total_tokens": handler.total_tokens
            },
            tool_use_blocks=[]
        )

    def _stream_native(
        self,
        messages: list[LLMMessage],
        system: str | None = None
    ) -> Iterator[str]:
        """
        Нативный streaming от провайдера.

        Args:
            messages: История сообщений
            system: Системный промпт

        Yields:
            Токены
        """
        # Это заглушка - реальная реализация зависит от провайдера
        # Для Ollama, OpenAI, Anthropic нужны разные реализации

        provider = self.llm_client.config.provider.value

        if provider == "ollama":
            yield from self._stream_ollama(messages, system)
        elif provider == "openai":
            yield from self._stream_openai(messages, system)
        elif provider == "anthropic":
            yield from self._stream_anthropic(messages, system)
        else:
            # Fallback - эмуляция
            response = self.llm_client.complete(messages, system)
            yield from self._chunk_text(response.content)

    def _stream_ollama(
        self,
        messages: list[LLMMessage],
        system: str | None = None
    ) -> Iterator[str]:
        """Streaming для Ollama."""
        import requests

        url = f"{self.llm_client.config.base_url}/api/chat"

        payload = {
            "model": self.llm_client.config.model,
            "messages": [{"role": m.role, "content": m.content} for m in messages],
            "stream": True
        }

        if system:
            payload["system"] = system

        try:
            response = requests.post(url, json=payload, stream=True, timeout=120)
            response.raise_for_status()

            for line in response.iter_lines():
                if line:
                    import json
                    data = json.loads(line)
                    if "message" in data and "content" in data["message"]:
                        yield data["message"]["content"]

        except Exception as e:
            print(f"Streaming error: {e}")
            return

    def _stream_openai(
        self,
        messages: list[LLMMessage],
        system: str | None = None
    ) -> Iterator[str]:
        """Streaming для OpenAI."""
        # Заглушка - нужна реализация с openai SDK
        response = self.llm_client.complete(messages, system)
        yield from self._chunk_text(response.content)

    def _stream_anthropic(
        self,
        messages: list[LLMMessage],
        system: str | None = None
    ) -> Iterator[str]:
        """Streaming для Anthropic."""
        # Заглушка - нужна реализация с anthropic SDK
        response = self.llm_client.complete(messages, system)
        yield from self._chunk_text(response.content)

    def _emulate_streaming(self, text: str, handler: StreamingHandler) -> None:
        """
        Эмулировать streaming для провайдеров без поддержки.

        Args:
            text: Полный текст
            handler: Обработчик токенов
        """
        # Разбиваем на слова и выводим с задержкой
        words = text.split()
        for i, word in enumerate(words):
            token = word + (" " if i < len(words) - 1 else "")
            handler.handle_token(token)
            time.sleep(0.02)  # 20ms задержка между словами

    def _chunk_text(self, text: str, chunk_size: int = 5) -> Iterator[str]:
        """
        Разбить текст на чанки для эмуляции streaming.

        Args:
            text: Текст
            chunk_size: Размер чанка в словах

        Yields:
            Чанки текста
        """
        words = text.split()
        for i in range(0, len(words), chunk_size):
            chunk = " ".join(words[i:i + chunk_size])
            if i + chunk_size < len(words):
                chunk += " "
            yield chunk
            time.sleep(0.05)  # Небольшая задержка


class StreamingREPL:
    """REPL с поддержкой streaming ответов."""

    def __init__(self, streaming_client: StreamingLLMClient):
        self.streaming_client = streaming_client
        self.messages: list[LLMMessage] = []

    def process_message(self, user_message: str, system: str | None = None) -> str:
        """
        Обработать сообщение с streaming.

        Args:
            user_message: Сообщение пользователя
            system: Системный промпт

        Returns:
            Полный ответ
        """
        # Добавляем сообщение пользователя
        self.messages.append(LLMMessage(role="user", content=user_message))

        print("\n🤖 Assistant: ", end="", flush=True)

        # Получаем ответ с streaming
        response = self.streaming_client.complete_streaming(
            messages=self.messages,
            system=system
        )

        print("\n")  # Новая строка после ответа

        # Добавляем ответ в историю
        self.messages.append(LLMMessage(role="assistant", content=response.content))

        return response.content


# Пример использования
if __name__ == "__main__":
    from .llm_client import LLMConfig, LLMProvider, create_llm_client

    # Создаем клиента
    config = LLMConfig(
        provider=LLMProvider.OLLAMA,
        model="llama3.2:1b",
        api_key=""
    )

    llm_client = create_llm_client(config)
    streaming_client = StreamingLLMClient(llm_client)

    # Тестируем streaming
    messages = [
        LLMMessage(role="user", content="Tell me a short story about a robot")
    ]

    print("Testing streaming response:\n")
    response = streaming_client.complete_streaming(messages)
    print(f"\n\nTotal tokens: {response.usage.get('completion_tokens', 0)}")

File: src/streaming.rs
//! Streaming Response — потоковая передача ответа LLM по мере генерации.
//! Порт `streaming.py` (итерация 6). Как и в оригинале, честно
//! реализован НАСТОЯЩИЙ streaming только для Ollama (единственный
//! провайдер, для которого `streaming.py` тоже не был заглушкой —
//! `_stream_openai`/`_stream_anthropic` в оригинале сами были no-op
//! обёртками вокруг обычного `complete()` с последующей псевдо-разбивкой
//! на чанки). Для всех остальных провайдеров здесь — то же самое:
//! обычный `LLMClient::complete()`, результат режется на чанки и
//! отдаётся вызывающему коду с искусственной задержкой, чтобы UX
//! (постепенный вывод текста в REPL) выглядел одинаково независимо от
//! провайдера.
//!
//! НЕ подключено в основной цикл `Agent::chat()` в `main.rs` — тот
//! всегда использует `complete_with_tools()` (нужно для нативного tool
//! calling), а стриминг токен-за-токеном и параллельное выполнение
//! инструментов из недописанного потока — взаимоисключающие вещи без
//! существенного изменения основного цикла (см. также
//! `streaming_tool_calling.rs`, который решает именно эту задачу для
//! моделей без API tool-calling, но тоже не встроен в главный цикл —
//! по той же причине, что и `parser.rs`, см. STATUS_RU.md).

use futures_util::StreamExt;
use std::time::Duration;

use crate::llm::{LLMClient, LLMMessage};

/// Потоково стримит ответ Ollama по HTTP (NDJSON-чанки), вызывая
/// `on_token` для каждого куска текста. Возвращает полный склеенный
/// ответ. При любой ошибке сети — тихий fallback на `stream_emulated`.
pub async fn stream_ollama<F: FnMut(&str)>(
    base_url: &str,
    model: &str,
    messages: &[LLMMessage],
    system: Option<&str>,
    mut on_token: F,
) -> String {
    let client = reqwest::Client::new();
    let url = format!("{}/api/chat", base_url);

    let mut ollama_messages: Vec<serde_json::Value> = Vec::new();
    if let Some(s) = system {
        ollama_messages.push(serde_json::json!({"role": "system", "content": s}));
    }
    for m in messages {
        ollama_messages.push(serde_json::json!({"role": m.role, "content": m.content}));
    }

    let body = serde_json::json!({
        "model": model,
        "messages": ollama_messages,
        "stream": true,
    });

    let resp = match client.post(&url).json(&body).send().await {
        Ok(r) if r.status().is_success() => r,
        _ => return String::new(),
    };

    let mut full = String::new();
    let mut byte_stream = resp.bytes_stream();
    let mut leftover = String::new();

    while let Some(chunk) = byte_stream.next().await {
        let Ok(bytes) = chunk else { break };
        leftover.push_str(&String::from_utf8_lossy(&bytes));

        // Ollama шлёт NDJSON — один JSON-объект на строку.
        while let Some(pos) = leftover.find('\n') {
            let line = leftover[..pos].trim().to_string();
            leftover.drain(..=pos);
            if line.is_empty() {
                continue;
            }
            if let Ok(data) = serde_json::from_str::<serde_json::Value>(&line) {
                if let Some(token) = data["message"]["content"].as_str() {
                    if !token.is_empty() {
                        on_token(token);
                        full.push_str(token);
                    }
                }
                if data["done"].as_bool() == Some(true) {
                    return full;
                }
            }
        }
    }
    full
}

/// Разбить готовый текст на чанки по `chunk_size` слов и "проиграть" их
/// вызывающему коду с искусственной задержкой — эмуляция streaming для
/// провайдеров без нативной потоковой выдачи (см. заголовок файла).
pub async fn stream_emulated<F: FnMut(&str)>(text: &str, chunk_size: usize, mut on_chunk: F) {
    let words: Vec<&str> = text.split_whitespace().collect();
    if words.is_empty() {
        return;
    }
    let mut i = 0;
    while i < words.len() {
        let end = (i + chunk_size).min(words.len());
        let mut chunk = words[i..end].join(" ");
        if end < words.len() {
            chunk.push(' ');
        }
        on_chunk(&chunk);
        tokio::time::sleep(Duration::from_millis(50)).await;
        i = end;
    }
}

/// Получить ответ LLM с потоковым выводом. Для Ollama — настоящий
/// streaming, для остальных провайдеров — обычный `complete()` +
/// пост-фактум эмуляция чанков (см. заголовок файла — так делал и
/// оригинал).
pub async fn complete_streaming<F: FnMut(&str)>(
    client: &dyn LLMClient,
    provider_name: &str,
    base_url: Option<&str>,
    model: &str,
    messages: &[LLMMessage],
    system: Option<&str>,
    mut on_token: F,
) -> Result<String, crate::llm::LLMError> {
    if provider_name == "ollama" {
        let url = base_url.unwrap_or("http://localhost:11434");
        let text = stream_ollama(url, model, messages, system, &mut on_token).await;
        if !text.is_empty() {
            return Ok(text);
        }
        // Streaming не удался (сеть/парсинг) — падаем на обычный запрос.
    }

    let response = client.complete(messages, system).await?;
    stream_emulated(&response.content, 5, &mut on_token).await;
    Ok(response.content)
}
